### PR TITLE
magento/devdocs#: Expected exceptions

### PR DESCRIPTION
**Pull request by atwixfirster**
## This PR is a:

- [ ] New topic
- [X] Content update
- [ ] Content fix or rewrite
- [ ] Bug fix or improvement

## Summary

When this pull request is merged, it will add information to "GraphQL functional testing" page how to cover negative scenarios with expected exceptions. 

## Additional information

List all affected URLs 

- https://devdocs.magento.com/guides/v2.3/graphql/functional-testing.html

Thank you!

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.3/graphql/functional-testing.md [hunk 1]
```diff
@@ -170,3 +170,123 @@ $registry->register('isSecureArea', false);
 
 A fixture rollback filename should correspond to the original fixture filename postfixed by `_rollback` keyword. (For example, `virtual_product_rollback.php`.) 
 Magento provides fixtures in the `dev/tests/integration/testsuite/Magento/<ModuleName>/_files` directory. Use these fixtures whenever possible. When you create your own fixture, also create a proper rollback.
+
+## Expected exceptions
+Sometimes we need to cover a negative scenarious in our test and make sure that our functionality proper responses on them. There are two ways how you can define the expected exception message:
```
Comment (rogyar) [inline guides/v2.3/graphql/functional-testing.md]: Please, rephrase the description slightly by removing "we, our etc" occurrences. In case of documentation, such writing style might be a bit confusing. In result, it will be something like: "In cases when it's necessary to cover negative scenarios ... There are two ways how the expected exception message can be defined".

File: guides/v2.3/graphql/functional-testing.md [hunk 2]
```diff
@@ -170,3 +170,123 @@ $registry->register('isSecureArea', false);
...
+        );
+        $maskedQuoteId = $this->quoteIdToMaskedId->execute((int)$this->quote->getId());
+        $query = $this->prepareGetCartQuery($maskedQuoteId);
+        self::expectExceptionMessage("The current user cannot perform operations on cart \"$maskedQuoteId\"");
+        $this->graphQlQuery($query);
+    } 
+```
+
+#### Define expected exception message in test function annotation
+```php
```
Comment (rogyar) [inline guides/v2.3/graphql/functional-testing.md]: Please add a sentence after the header that says: "As an alternative for calling `expectExceptionMessage` method" a predefined directive `@expectedExceptionMessage` can be used for a test

File: guides/v2.3/graphql/functional-testing.md [hunk 3]
```diff
@@ -170,3 +170,123 @@ $registry->register('isSecureArea', false);
...
+    /**
+     * @expectedException \Exception
+     * @expectedExceptionMessage Expected exception message goes here...
+     */
+```
+
+Let's imagine that customer wants to retrieve information about his own cart but he provides incorrect cart ID.
```
Comment (rogyar) [inline guides/v2.3/graphql/functional-testing.md]: Please, replace "Let's imagine" with "Example: a customer wants ... "

File: guides/v2.3/graphql/functional-testing.md [hunk 4]
```diff
@@ -170,3 +170,123 @@ $registry->register('isSecureArea', false);
 
 A fixture rollback filename should correspond to the original fixture filename postfixed by `_rollback` keyword. (For example, `virtual_product_rollback.php`.) 
 Magento provides fixtures in the `dev/tests/integration/testsuite/Magento/<ModuleName>/_files` directory. Use these fixtures whenever possible. When you create your own fixture, also create a proper rollback.
+
+## Expected exceptions
+Sometimes we need to cover a negative scenarious in our test and make sure that our functionality proper responses on them. There are two ways how you can define the expected exception message:
+  - in the body of test function
+  - in the test function annotation 
+
+#### Define exception message in the body of test
```
Comment (rogyar) [inline guides/v2.3/graphql/functional-testing.md]: Please, replace with "Define exception message expectation ... "

File: guides/v2.3/graphql/functional-testing.md [hunk 5]
```diff
@@ -170,3 +170,123 @@ $registry->register('isSecureArea', false);
...
+            $this->quote,
+            $reservedOrderId,
+            'reserved_order_id'
+        );
+        $maskedQuoteId = $this->quoteIdToMaskedId->execute((int)$this->quote->getId());
+        $query = $this->prepareGetCartQuery($maskedQuoteId);
+        self::expectExceptionMessage("The current user cannot perform operations on cart \"$maskedQuoteId\"");
+        $this->graphQlQuery($query);
+    } 
+```
```
Comment (rogyar) [inline guides/v2.3/graphql/functional-testing.md]: I would also add something like:
"Please note that the exception message expectation should be defined before invoking a logic that generates the exception"

**Comment by rogyar** [comment]
Hi @atwixfirster. Thank you for your collaboration. Please, consider making minor adjustments mentioned in the review.

File: guides/v2.3/graphql/functional-testing.md (hunk 5 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

@rogyar , could you please review?

File: guides/v2.3/graphql/functional-testing.md (hunk 4 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

File: guides/v2.3/graphql/functional-testing.md (hunk 3 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

File: guides/v2.3/graphql/functional-testing.md (hunk 2 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

File: guides/v2.3/graphql/functional-testing.md (hunk 1 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

Comment (shrielenee) [reply guides/v2.3/graphql/functional-testing.md]: Or to minimize further, it could simply be.. "Sometimes you need to cover a negative scenario in your test to ensure..."

It's okay to use "you". Just keep the sentences simple and concise.

Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: Hi, @shrielenee ! Thank you for your notice.

Could you please clarify if this version is ok?

**Current version:** 
```In cases when it's necessary to cover negative scenarios in the test and make sure that testing functionality proper responses on them use the next ways to define the expected exception message:```

Thanks!

Comment (shrielenee) [reply guides/v2.3/graphql/functional-testing.md]: I don't understand what this phrase means --- "make sure that testing functionality proper responses on them"

I would reword it similarly to...

> When it is necessary to cover negative test scenarios and ensure that testing elicits proper responses, define the expected exception message:

Comment (shrielenee) [reply guides/v2.3/graphql/functional-testing.md]: But this PR needs a copyedit for language/grammar when the technical content is done.

Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: > I would reword it similarly to...

**implemented**

Thank you, @shrielenee !


> But this PR needs a copyedit for language/grammar when the technical content is done.

@rogyar , do you have any suggestions to the technical content?

Many thanks!

Comment (shrielenee) [reply guides/v2.3/graphql/functional-testing.md]: Ping me when technical content is done and I'll copyedit.

**Comment by keharper** [comment]
I've got this, @shrielenee . This PR expands https://github.com/magento/devdocs/pull/4082/

**Comment by keharper** [comment]
@atwixfirster  please pull in the changes from this PR

https://github.com/magento/devdocs/pull/4082

File: guides/v2.3/graphql/functional-testing.md [hunk 6]
```diff
@@ -170,3 +170,126 @@ $registry->register('isSecureArea', false);
 
 A fixture rollback filename should correspond to the original fixture filename postfixed by `_rollback` keyword. (For example, `virtual_product_rollback.php`.) 
 Magento provides fixtures in the `dev/tests/integration/testsuite/Magento/<ModuleName>/_files` directory. Use these fixtures whenever possible. When you create your own fixture, also create a proper rollback.
+
+## Expected exceptions
+When it is necessary to cover negative test scenarios and ensure that testing elicits proper responses, define the expected exception message:
+  - in the body of test
+  - in test function annotation
+
+#### Define exception message expectation in the body of test
```
Comment (keharper) [inline guides/v2.3/graphql/functional-testing.md]: Delete one of the #s. The previous heading was a H2, so this one should be a H3.

File: guides/v2.3/graphql/functional-testing.md [hunk 7]
```diff
@@ -170,3 +170,126 @@ $registry->register('isSecureArea', false);
...
+            'reserved_order_id'
+        );
+        $maskedQuoteId = $this->quoteIdToMaskedId->execute((int)$this->quote->getId());
+        $query = $this->prepareGetCartQuery($maskedQuoteId);
+        self::expectExceptionMessage("The current user cannot perform operations on cart \"$maskedQuoteId\"");
+        $this->graphQlQuery($query);
+    } 
+```
+
+#### Define expected exception message in test function annotation
```
Comment (keharper) [inline guides/v2.3/graphql/functional-testing.md]: Change this to a H3 as well

File: guides/v2.3/graphql/functional-testing.md [hunk 8]
```diff
@@ -170,3 +170,126 @@ $registry->register('isSecureArea', false);
...
+    public function testGetNonExistentCart()
+    {
+        $maskedQuoteId = 'non_existent_masked_id';
+        $query = $this->prepareGetCartQuery($maskedQuoteId);
+        
+        $this->graphQlQuery($query);
+    }
+```
+
+There is a list of functions which you can use to cover expected exceptions:
```
Comment (keharper) [inline guides/v2.3/graphql/functional-testing.md]:
```suggestion
Use the following functions to cover expected exceptions:  
```

File: guides/v2.3/graphql/functional-testing.md [hunk 9]
```diff
@@ -170,3 +170,126 @@ $registry->register('isSecureArea', false);
...
+    {
+        $maskedQuoteId = 'non_existent_masked_id';
+        $query = $this->prepareGetCartQuery($maskedQuoteId);
+        
+        $this->graphQlQuery($query);
+    }
+```
+
+There is a list of functions which you can use to cover expected exceptions:  
+- expectException
```
Comment (keharper) [inline guides/v2.3/graphql/functional-testing.md]: Wrap the functions with a text code block

File: guides/v2.3/graphql/functional-testing.md (hunk 8 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: > Suggested change
> There is a list of functions which you can use to cover expected exceptions:  
> Use the following functions to cover expected exceptions:

**done**

File: guides/v2.3/graphql/functional-testing.md (hunk 9 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: > Wrap the functions with a text code block

**done**

Thank you, @keharper 👍

File: guides/v2.3/graphql/functional-testing.md (hunk 7 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **done**

File: guides/v2.3/graphql/functional-testing.md (hunk 6 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **done**

**Comment by atwixfirster** [comment]
> @atwixfirster please pull in the changes from this PR
> 
> #4082

**done**

File: guides/v2.3/graphql/functional-testing.md [hunk 10]
```diff
@@ -193,3 +193,128 @@ try {
...
+            'reserved_order_id'
+        );
+        $maskedQuoteId = $this->quoteIdToMaskedId->execute((int)$this->quote->getId());
+        $query = $this->prepareGetCartQuery($maskedQuoteId);
+        self::expectExceptionMessage("The current user cannot perform operations on cart \"$maskedQuoteId\"");
+        $this->graphQlQuery($query);
+    } 
+```
+
+### Define expected exception message in test function annotation
```
Comment (keharper) [inline guides/v2.3/graphql/functional-testing.md]:
```suggestion
### Exception messages in the annotation of a test function
```

File: guides/v2.3/graphql/functional-testing.md [hunk 11]
```diff
@@ -193,3 +193,128 @@ try {
...
+As an alternative for calling `expectExceptionMessage` method use a predefined directive `@expectedExceptionMessage`:
+
+```php
+    /**
+     * @expectedException \Exception
+     * @expectedExceptionMessage Expected exception message goes here...
+     */
+```
+
+Example: a customer wants to retrieve information about his own cart but he provides incorrect cart ID.
```
Comment (keharper) [inline guides/v2.3/graphql/functional-testing.md]:
```suggestion
In the following query, a customer provides an incorrect cart ID while trying to retrieve information about his own cart.
```

File: guides/v2.3/graphql/functional-testing.md [hunk 12]
```diff
@@ -193,3 +193,128 @@ try {
...
+      ]
+    }
+  ],
+  "data": {
+    "cart": null
+  }
+}
+```
+
+See the code below how to cover this example with `@expectExceptionMessage` annotation:
```
Comment (keharper) [inline guides/v2.3/graphql/functional-testing.md]:
```suggestion
The `@expectExceptionMessage` annotation provides the text for the exception in this test.
```

File: guides/v2.3/graphql/functional-testing.md [hunk 13]
```diff
@@ -193,3 +193,128 @@ try {
 $registry->unregister('isSecureArea');
 $registry->register('isSecureArea', false);
 ```
+
+## Expected exceptions
+When it is necessary to cover negative test scenarios and ensure that testing elicits proper responses, define the expected exception message:
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **implemented**

File: guides/v2.3/graphql/functional-testing.md [hunk 14]
```diff
@@ -193,3 +193,128 @@ try {
...
+  - in the body of test
+  - in test function annotation
+
+### Define exception message expectation in the body of test
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

File: guides/v2.3/graphql/functional-testing.md [hunk 15]
```diff
@@ -193,3 +193,128 @@ try {
...
+
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **done**

Could you please review?

File: guides/v2.3/graphql/functional-testing.md [hunk 16]
```diff
@@ -193,3 +193,128 @@ try {
...
+    ...
+    
+    $this->expectExceptionMessage("Expected exception message goes here...");
+    
+    ...
+}
+```
+
+{:.bs-callout .bs-callout-info}
+Please note that the exception message expectation should be defined before invoking a logic that generates the exception.
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **implemented**

File: guides/v2.3/graphql/functional-testing.md [hunk 17]
```diff
@@ -193,3 +193,128 @@ try {
...
+
+For example, if `customer A` wants to retrieve information about `customer B cart. In this situation, `customer A` will get an error - "`The current user cannot perform operations on cart "XXXXX`", where `XXXXX` - is an unique ID of `customer B` cart. See code below to understand how to cover this example with `expectExceptionMessage` function:
```
[the remark this answers is not in the document]
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **implemented**

File: guides/v2.3/graphql/functional-testing.md (hunk 10 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

File: guides/v2.3/graphql/functional-testing.md (hunk 11 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

File: guides/v2.3/graphql/functional-testing.md (hunk 12 above)
Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: **fixed**

Comment (atwixfirster) [reply guides/v2.3/graphql/functional-testing.md]: Thank you very much for your review , @keharper !

**Comment by keharper** [comment]
I built this file locally and saw that some changes to the spacing need to be made. So I did that, made a few grammatical changes, and also added Atwix as the author, since more than half of the content was contributed by your company in the last couple of weeks.

**Comment by keharper** [comment]
running tests

**Comment by keharper** [comment]
Thanks for your contribution @atwixfirster

**Comment by dshevtsov** [comment]
@keharper the PR contains two labels to indicate the type of changes: Technical and Major update. Please use only one and add whatsnew if needed.

**Comment by dshevtsov** [comment]
@keharper add labels to indicate the affected versions.
